File: packages_ts/backend/archivist/src/general-search/general-search.service.ts
import {
  textSearchQuery,
  countTextSearchQuery,
  uidSearchQuery,
  countUIDSearchQuery,
} from '../graph/queries.js';
import { Injectable } from '@nestjs/common';
import { ExecuteSearchQueryService } from '../execute-search-query/execute-search-query.service.js';
import { CacheService } from '../cache/cache.service.js';

@Injectable()
export class GeneralSearchService {
  constructor(
    private readonly executeSearchQueryService: ExecuteSearchQueryService,
    private readonly cacheService: CacheService,
  ) {}

  async getTextSearch(
    searchTerm: string,
    collectionUID: number,
    page = 1,
    pageSize = 10,
    filter = null,
    exactMatch = false,
  ) {
    let descendants = [];
    if (filter) {
      descendants = await this.cacheService.allDescendantsOf(
        parseInt(filter.uid),
      );
    }
    const res = await this.executeSearchQueryService.executeSearchQuery(
      textSearchQuery,
      countTextSearchQuery,
      searchTerm,
      [1146, 1726, 1981, 1225],
      descendants,
      collectionUID,
      page,
      pageSize,
      exactMatch,
    );
    console.log('TEXT SEARCH: ', res);
    return res;
  }

  async getUIDSearch(
    searchTerm: number,
    collectionUID: number,
    page = 1,
    pageSize = 10,
    filter = null,
  ) {
    let descendants = [];
    if (filter) {
      descendants = await this.cacheService.allDescendantsOf(
        parseInt(filter.uid),
      );
    }
    const res = await this.executeSearchQueryService.executeSearchQuery(
      uidSearchQuery,
      countUIDSearchQuery,
      searchTerm,
      [1146, 1726, 1981, 1225],
      descendants,
      collectionUID,
      page,
      pageSize,
    );
    return res;
  }
}

File: packages_clj/archivist/src/io/relica/archivist/services/general_search_service.clj
(ns io.relica.archivist.services.general-search-service
  (:require [clojure.tools.logging :as log]
            [clojure.core.async :refer [<! go]]
            [io.relica.archivist.services.graph-service :as graph]
            [io.relica.archivist.services.cache-service :as cache]
            [io.relica.archivist.db.queries :as queries]))

(defprotocol GeneralSearchOperations
  (get-text-search [this search-term collection-uid page page-size filter exact-match])
  (get-uid-search [this search-term collection-uid page page-size filter]))

(defrecord GeneralSearchService [graph-service cache-service]
  GeneralSearchOperations
  
  (get-text-search [_ search-term collection-uid page page-size filter exact-match]
    (go
      (try
        (let [descendants (when filter
                           (<! (cache/all-descendants-of cache-service (:uid filter))))
              rel-type-uids [1146 1726 1981 1225]
              skip (* (dec page) page-size)
              results (graph/exec-query graph-service
                                      queries/text-search
                                      {:searchTerm search-term
                                       :relTypeUIDs rel-type-uids
                                       :filterUIDs (or descendants [])
                                       :collectionUID (or collection-uid "")
                                       :exactMatch exact-match
                                       :skip skip
                                       :pageSize page-size})
              count-results (graph/exec-query graph-service
                                            queries/count-text-search
                                            {:searchTerm search-term
                                             :relTypeUIDs rel-type-uids
                                             :filterUIDs (or descendants [])
                                             :collectionUID (or collection-uid "")
                                             :exactMatch exact-match})
              total (get-in (first count-results) [:total])]
          {:totalCount total
           :page page
           :pageSize page-size
           :facts (graph/transform-results results)})
        (catch Exception e
          (log/error e "Error in text search")
          {:error "Failed to execute text search"}))))

  (get-uid-search [_ search-term collection-uid page page-size filter]
    (go
      (try
        (let [descendants (when filter
                           (<! (cache/all-descendants-of cache-service (:uid filter))))
              rel-type-uids [1146 1726 1981 1225]
              skip (* (dec page) page-size)
              results (graph/exec-query graph-service
                                      queries/uid-search
                                      {:searchTerm search-term
                                       :relTypeUIDs rel-type-uids
                                       :filterUIDs (or descendants [])
                                       :collectionUID (or collection-uid "")
                                       :skip skip
                                       :pageSize page-size})
              count-results (graph/exec-query graph-service
                                            queries/count-uid-search
                                            {:searchTerm search-term
                                             :relTypeUIDs rel-type-uids
                                             :filterUIDs (or descendants [])
                                             :collectionUID (or collection-uid "")})
              total (get-in (first count-results) [:total])]
          {:totalCount total
           :page page
           :pageSize page-size
           :facts (graph/transform-results results)})
        (catch Exception e
          (log/error e "Error in UID search")
          {:error "Failed to execute UID search"})))))

(defn create-general-search-service [{:keys [graph
                                             cache]}]
  (->GeneralSearchService graph cache))

(defonce general-search-service (atom nil))

(defn start [services]
  (println "Starting General Search Service...")
  (let [service (create-general-search-service services)]
    (reset! general-search-service service)
    service))

(defn stop []
  (println "Stopping General Search Service..."))

(comment
  ;; Test operations
  @general-search-service
  ;;(get-text-search [_ search-term collection-uid page page-size filter exact-match]
  (go
    (let [results (<! (get-text-search @general-search-service "r" nil 1 10 nil false))]
    (tap> results)
    (println results)))

  )
